nemo/collections/nlp/modules/common/text_generation_strategy.py

```diff
@@ -280,18 +285,118 @@ def __init__(self, model, **args):
         super().__init__(model)
         self.forward_model = self.model.model
         self.frequent_query = args['frequent_query']
-        del args['frequent_query']
+        self.pad_token_for_retrieval = args['pad_tokens']
         self.neighbors = args['neighbors']
-        self.service = FaissRetrievalService(tokenizer=self.model.tokenizer, **args)
+        self.store_retrieved = args['store_retrieved']
+        weights = args['weights']
+        # start the sentence bert server
+        start_sentence_bert_server(tokenizer=self.model.tokenizer, **args['sentence_bert'])
+        services = []
+        for service_conf in args['services']:
+            if service_conf['type'] == 'FaissRetrievalService':
+                del service_conf['type']
+                service = FaissRetrievalService(tokenizer=self.model.tokenizer, **service_conf)
+                services.append(service)
+            elif service_conf['type'] == 'DynamicFaissRetrievalService':
+                del service_conf['type']
+                service = DynamicFaissRetrievalService(tokenizer=self.model.tokenizer, **service_conf)
+                services.append(service)
+            else:
+                raise ValueError(f'no such service {service_conf["type"]} implemented')
+        self.service = ComboRetrievalService(retrieval_services=services, weights=weights)
+        self.retrieved = []
+        self.retrieved_text = []
+        self.chunk_size = self.service.chunk_size
+
+    def update_neighbors(self, neighbors):
+        # dynamically change the number of neighbors during the query
+        self.neighbors = neighbors
+
+    def update_weights(self, weights):
+        # dynamically change the weights between different retrieval services
+        self.service.update_weights(weights)
+
+    def tokenize_batch(self, sentences, max_len, add_BOS):
+        """
+        convert the sentences into lists of tokens, pad them to the same length, add bos tokens if it is needed
+        Args:
+            sentences (List[str]): list of input sentences in str format.
+            max_len (int): max number of tokens to generate.
+            add_BOS (bool): whether to add the BOS token at the beginning
+        Returns:
+            Tuple[torch.Tensor], the tokenized and padded torch tensor and the token context length tensor.
+        """
+        tokenizer = self.model.tokenizer
+        if add_BOS:
```

Comment:
For `add_BOS` you are adding `tokenizer.eos_id` here. Is this corrrect?

Reply:
yeah. the GPT tokenizer uses "eos" as "bos". tokenizer doesn't distinguish them. But you are right, bos_id is better here in case other tokenizer uses a different id. fixed